### PR TITLE
Disable output file map diagnostic for non-incremental mode

### PR DESCRIPTION
**Pull request by keith**
This re-implements this logic https://github.com/apple/swift/blob/3116eed2e465a2688f070b3b87adc106cbbaf09f/lib/Driver/Driver.cpp#L414-L417 to not warn when output file maps don't have a main entry when using WMO

### CODE REVIEW

**Review by brentleyjones — APPROVED** [review]

**Review by davidungar — CHANGES_REQUESTED** [review]
Thank you for doing this. It looks good, but before merging I'd like to see an XCTest added to check this. It might be easiest to add a test in `IncrementalCompilationTests.swift`, in which the test runs the driver both with- and without the `-incremental` flag and checks to see whether the diagnostic is emitted when appropriate.

Would you like to add such a test?

**Comment by brentleyjones** [comment]
FYI, this is still an issue in Xcode 13.0 beta 4.

**Comment by keith** [comment]
Updated with tests!

**Comment by artemcm** [comment]
@swift-ci please test

**Review by artemcm — APPROVED** [review]

**Comment by artemcm** [comment]
@davidungar could you please take another look? It looks good to go to me.

**Comment by keith** [comment]
@davidungar mind taking another look here?

**Comment by keith** [comment]
@davidungar mind taking another look here?

File: Tests/TestUtilities/OutputFileMapCreator.swift [hunk 1]
```diff
@@ -18,23 +18,27 @@ public struct OutputFileMapCreator {
   private let module: String
   private let inputPaths: [AbsolutePath]
   private let derivedData: AbsolutePath
+  private let excludeMainEntry: Bool
```
Comment (davidungar) [inline Tests/TestUtilities/OutputFileMapCreator.swift:21]: Would it be possible to add a comment somewhere explaining when you would want to exclude the main entry?

**Review by davidungar — APPROVED** [review]
Sorry this PR slipped under my radar. Thanks to @CodaFi for pinging me. In future, please feel free to ping me directly. 
It would be great if you could add a comment somewhere (if it's not already there) explaining why one would want to exclude the main entry, or under what circumstances. Otherwise, LGTM and thank you for your patience!

**Comment by keith** [comment]
Done, thanks for following up. To be clear where do you mean when you say "ping you directly"?

**Review by artemcm — APPROVED** [review]

**Comment by artemcm** [comment]
@swift-ci please test

**Comment by brentleyjones** [comment]
@artemcm Is this able to get CP'ed into 5.5?

**Comment by artemcm** [comment]
> @artemcm Is this able to get CP'ed into 5.5?

Yeah, I think this change is pretty safe.
Please create a cherry-pick PR against `release/5.5` and use something like https://github.com/apple/swift/pull/39140 as a template for the PR description.

**Comment by keith** [comment]
Thanks! Submitted https://github.com/apple/swift-driver/pull/828 for the CP

**Pull request merged**